app/ledger.js

```diff
@@ -216,6 +234,51 @@ var boot = () => {
 }
 
 /*
+ * Print or Save Recovery Keys
+ */
+
+var backupKeys = (appState, action) => {
+  const paymentId = appState.getIn(['ledgerInfo', 'paymentId'])
+  const passphrase = appState.getIn(['ledgerInfo', 'passphrase'])
+  const message = locale.translation('ledgerBackupText', {paymentId, passphrase})
+  const filePath = path.join(app.getPath('userData'), '/brave_wallet_recovery.txt')
+
+  fs.writeFile(filePath, message, (err) => {
+    if (err) {
+      console.log(err)
+    } else {
+      Tabs.create({url: fileUrl(filePath)}).then((webContents) => {
+        if (action.backupAction === 'print') {
+          webContents.print({silent: false, printBackground: false})
+        } else {
+          webContents.downloadURL(fileUrl(filePath))
+        }
+      }).catch((err) => {
+        console.error(err)
+      })
+    }
+  })
+
+  return appState
+}
+
+/*
+ * Recover Ledger Keys
+ */
+
+var recoverKeys = (appState, action) => {
+  client.recoverWallet(action.firstRecoveryKey, action.secondRecoveryKey, (err, body) => {
```

Comment:
this callback may or may not happen asynchronously so the appActions should be called inside `setImmediate` to prevent any potential race conditions with state updates. The issue was discussed in slack, but this is how it happens:

```
For simplicity, all actions increment the current count by 1, but the same problem exists even if the actions change different parts of the state
1. action1 is called with the current appState = {count: 0}
2. action1 calls a method with a callback
3. callback returns synchronously (does not perform any io or other operations that would make it async)
4. action2 is called inside callback with the current appState = {count: 0}
5. action2 returns appState = {count: 1}
6. appState is emitted
7. action1 returns appState = {count: 1}
8 app state is emitted
```